extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java

```diff
@@ -221,28 +228,105 @@ void recordBouncyCastleProviders(SecurityProviderRecorder recorder,
-                    : SecurityProviderUtils.BOUNCYCASTLE_PROVIDER_CLASS_NAME;
-            additionalProviders.produce(new NativeImageSecurityProviderBuildItem(providerName));
+
+            ClassCreator file = createBouncyCastleFeatureCreator(nativeImageClass);
+            MethodCreator beforeAn = createBeforeAnalysisMethod(file);
+            TryBlock overallCatch = beforeAn.tryBlock();
+
+            if (!bouncyCastleJsseProvider.get().isInFipsMode()) {
+                final int sunJsseIndex = findProviderIndex(SecurityProviderUtils.SUN_JSSE_PROVIDER_NAME);
+
+                ResultHandle bcProvider = overallCatch
+                        .newInstance(
+                                MethodDescriptor.ofConstructor(SecurityProviderUtils.BOUNCYCASTLE_PROVIDER_CLASS_NAME));
+                ResultHandle bcJsseProvider = overallCatch.newInstance(
+                        MethodDescriptor.ofConstructor(SecurityProviderUtils.BOUNCYCASTLE_JSSE_PROVIDER_CLASS_NAME));
+
+                overallCatch.invokeStaticMethod(
+                        MethodDescriptor.ofMethod(Security.class, "insertProviderAt", int.class, Provider.class,
+                                int.class),
+                        bcProvider, overallCatch.load(sunJsseIndex));
+                overallCatch.invokeStaticMethod(
+                        MethodDescriptor.ofMethod(Security.class, "insertProviderAt", int.class, Provider.class,
+                                int.class),
+                        bcJsseProvider, overallCatch.load(sunJsseIndex + 1));
+            } else {
+                final int sunIndex = findProviderIndex(SecurityProviderUtils.SUN_PROVIDER_NAME);
+
+                ResultHandle bcFipsProvider = overallCatch
+                        .newInstance(MethodDescriptor
+                                .ofConstructor(SecurityProviderUtils.BOUNCYCASTLE_FIPS_PROVIDER_CLASS_NAME));
+                MethodDescriptor bcJsseProviderConstructor = MethodDescriptor.ofConstructor(
+                        SecurityProviderUtils.BOUNCYCASTLE_JSSE_PROVIDER_CLASS_NAME, "boolean",
+                        Provider.class.getName());
+                ResultHandle bcJsseProvider = overallCatch.newInstance(bcJsseProviderConstructor,
+                        overallCatch.load(true), bcFipsProvider);
+
+                overallCatch.invokeStaticMethod(
+                        MethodDescriptor.ofMethod(Security.class, "insertProviderAt", int.class, Provider.class,
+                                int.class),
+                        bcFipsProvider, overallCatch.load(sunIndex));
+                overallCatch.invokeStaticMethod(
+                        MethodDescriptor.ofMethod(Security.class, "insertProviderAt", int.class, Provider.class,
+                                int.class),
+                        bcJsseProvider, overallCatch.load(sunIndex + 1));
+            }
+            completeBouncyCastleFeature(file, beforeAn, overallCatch);
```

Comment:
This is common code in both `if` and `else` blocks, I would move it after the if-statement to avoid code duplication.

This would also remove the need for defining the `completeBouncyCastleFeature` method.